proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,334 @@
+   to access the `EpochRewardProof`? What's the impact for solana downstream
+   project?
+
+#### Reward distribution challenges
+
+1. What are the incentives for the leader to pack reward redeem transactions?
+   What kind of block fees will the leader receive for packing such
+   transactions? Who will pay for the fees for such transactions? Who will be
+   the signer for those transactions? What prevents a malicious leader from not
+   packing those reward distribution transactions?
+
+   There are two ideas to incentivize the leader to include reward redeeming
+   transactions.
+
+   The first idea is to have the network reject the block that doesn't have all
+   the rewards redeemed. This would requires that all the nodes to keep track of
+   what rewards should be redeemed in a block, and verify that the actual block
+   have the expected rewards. Otherwise, the block will be rejected. The
+   downside of this idea is that it will add extra complexity in block replay
+   stage.
+
+   The second idea is to change up the inflation economics to add a "redeemer
+   reward". The validator can claim a portion of it from the merkle tree for
+   properly redeeming a set of stake rewards. For example, if there's 100 SOL in
+   rewards total, we add 1% of that for redeemer rewards, so 1 SOL. Each
+   redemption gives some proportion of the amount redeemed. If a validator
+   redeems 1000 lamports in a block, that's 10 lamports they can claim too. That
+   can even be done at the level of the stake program. This creates a race to
+   reward the biggest whales, but maybe that's ok?
+
+   Also, a quick math shows that for a stake to get 5000 lamports in rewards for
+   one epoch, it needs to be around 0.016 SOL (5000 lamports * 162.5 epochs per
+   year / 5% inflation / 1_000_000_000 lamports per SOL). So if we also restrict
+   the minimum reward earning amount to 1 SOL, and that earns 312,500 lamports /
+   epoch in rewards, with a minimum of 93k lamports at 1.5% inflation. 1% of
+   that is ~1k lamports, so not too far from a transaction fee. If they can pack
+   in 25 accounts in one transaction, hen they still come out on top.
+
+1. What's the performance impact for other transactions in the same block during
+   reward distribution? Will it impact the fee model for those blocks?
+
+1. Where should the `DistributeReward` instruction be executed? By the bank
+   runtime, or by some native program such as stake program, or by a specialized
+   bpf program in bpf executor? Which way is better?
+
+1. Will current transaction schedule be able to handle the new epoch reward
+   transactions? When there is a burst of transaction inflow for those blocks,
+   can we guarantee that the `DistributeReward` transaction will be scheduled?
+
+1.  How to handle validator restart and snapshot taking during reward
+    calculation/distribution?
+
+    - Option 1: prevent taking snapshot and restarting from snapshot during
```

Comment:
I feel it needs to some variant of Option 2. The cluster restart case in particular is problematic here because the current approach of using the `--wait-for-supermajority` flag requires the ledger/snapshot leading up to the restart slot already be present: the repair service in particular is not started until after the "`--wait-for-supermajority`" processing completes in the Labs validator.

Reply:
yes. I tried it on a test cluster. Option 1 doesn't work with cluster restart.

Another idea for option 2 is when we take snapshot for a bank during reward interval, we force to store the reward result. If the reward result is not available yet, we will wait till the RewardCalculationService have the result. This way, we don't need to store stake_cache in the snapshot. And there is only one kind of snapshot layout rather than two.